type-definitions/most.d.ts

```diff
@@ -1,385 +1,788 @@
-  b: Stream<B>,
-  c: Stream<C>
-): Stream<R>;
-export function zip<A, B, C, D, R>(
-  fn: (a: A, b: B, c: C, d: D) => R,
-  a: Stream<A>,
-  b: Stream<B>,
-  c: Stream<C>,
-  d: Stream<D>
-): Stream<R>;
-export function zip<A, B, C, D, E, R>(
-  fn: (a: A, b: B, c: C, d: D, e: E) => R,
-  a: Stream<A>,
-  b: Stream<B>,
-  c: Stream<C>,
-  d: Stream<D>,
-  e: Stream<E>
-): Stream<R>;
-
-export function recoverWith<A, B>(p: (a: B) => A, s: Stream<A>): Stream<A>;
-export function throwError(e: Error): Stream<any>;
-
-export function multicast<A>(s: Stream<A>): Stream<A>;
+declare module "most" {
+	type SeedValue<S, V> = { seed: S, value: V };
+	type TimeValue<V>    = { time: number, value: V };
+
+	// es5 does not have typings for this
+	// use any to allow type-casting
+	type Generator<A, B, C> = any
+	type Iterable = any;
+
+	type CreateGenerator<A> = (...args: Array<any>) => Generator<A|Promise<A>, any, any>;
+
+	export interface Sink<A> {
+		event(time: number, value: A): void;
+		end(time: number, value?: A): void;
+		error(time: number, err: Error): void;
+	}
+
+	export interface Task {
+		run(time: number): void;
+		error(time: number, e: Error): void;
+		dispose(): void;
+	}
+
+	export interface ScheduledTask {
+		task: Task;
+		run(): void;
+		error(err: Error): void;
+		dispose(): void;
+	}
+
+	export interface Timer {
+		now(): number;
+		setTimer(f: () => any, delayTime: number): Task | number;
+		clearTimer(task: Task | number): void;
+	}
+
+	export interface Timeline {
+		nextArrival(): number;
+		isEmpty(): boolean;
+		add(st: ScheduledTask): void;
+		remove(st: ScheduledTask): boolean;
+		removeAll(predicate: (st: Task) => boolean): void;
+		runTasks(time: number, runTask: (task: Task) => any): void;
+	}
+
+	export interface Scheduler {
+		now(): number;
+		asap(task: Task): ScheduledTask;
+		delay(delay: number, task: Task): ScheduledTask;
+		periodic(period: number, task: Task): ScheduledTask;
+		schedule(delay: number, period: number, task: Task): ScheduledTask;
+		cancel(task: ScheduledTask): void;
+		cancelAll(predicate: (task: Task) => boolean): void;
+	}
+
+	export interface Disposable<A> {
+		dispose(): void | Promise<A>;
+	}
+
+	export interface Source<A> {
+		run (sink: Sink<A>, scheduler: Scheduler): Disposable<A>;
+	}
+
+	export interface Subscriber<A> {
+		next(value: A): void;
+		error(err: Error): void;
+		complete(value?: A): void;
+	}
+
+	export interface Subscription<A> {
+		unsubscribe(): void;
+	}
+
+	export interface Stream<A> {
+		reduce<B>(f: (b: B, a: A) => B, b: B): Promise<B>;
+		observe(f: (a: A) => any): Promise<any>;
+		forEach(f: (a: A) => any): Promise<any>;
+		drain(): Promise<any>;
+		subscribe(subscriber: Subscriber<A>): Subscription<A>;
+
+		constant<B>(b: B): Stream<B>;
+		map<B>(f: (a: A) => B): Stream<B>;
+		tap(f: (a: A) => any): Stream<A>;
+		chain<B>(f: (a: A) => Stream<B>): Stream<B>;
+		flatMap<B>(f: (a: A) => Stream<B>): Stream<B>;
+
+		ap<B, C>(fs: Stream<(a: A) => B>): Stream<C>;
```

Comment:
This is confusing because Fantasy Land flipped the `ap` spec.  In the version of the spec (basically v0.3.0) that most.js 1.x supports, `ap`'s context (i.e. `this`) should be a Stream of functions, and it's argument should be a Stream of values.  Unfortunately, I think it's technically impossible to write an accurate type signature for it in TS or flow, which is side-benefit of the reversal in later Fantasy Land versions.

I guess the best we can do is what @brandonpayton suggested:

```ts
export interface Stream<A> {
    ap<B, C>(fs: Stream<B>): Stream<C>;
}
```

In most 2.0.0, we'll support Fantasy Land 2.1, which reverses `ap`, such that its possible to write a type signature for it.